kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -94,7 +94,7 @@ public ProduceAction(
       Provider<SchemaManager> schemaManagerProvider,
       Provider<RecordSerializer> recordSerializer,
       Provider<ProduceController> produceControllerProvider,
-      Provider<ProducerMetrics> producerMetrics,
+      ProducerMetrics producerMetrics,
```

Comment:
I'll admit I still don't 100% understand when to use Provider and when not to.

This change was based on it not working if I used a Provider, and also making our setup match what we do with producer https://github.com/confluentinc/kafka-rest/blob/5b869189dbdeb644b62560eff2233361bbb5f7b9/kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ProduceControllerImpl.java#L38

I think it's because the Provider makes the object too late, but I can't see why this would be a problem in the code below.

Let's talk to Rigel in the morning about this.

Reply:
For reference, this is the exception I get when I use a provider:

standalone-cloud-kafka-rest | java.lang.IllegalStateException: Not inside a request scope.
standalone-cloud-kafka-rest |   at org.glassfish.jersey.internal.guava.Preconditions.checkState(Preconditions.java:169)
standalone-cloud-kafka-rest |   at org.glassfish.jersey.process.internal.RequestScope.current(RequestScope.java:153)
standalone-cloud-kafka-rest |   at org.glassfish.jersey.inject.hk2.RequestContext.findOrCreate(RequestContext.java:55)
standalone-cloud-kafka-rest |   at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
standalone-cloud-kafka-rest |   at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:758)
standalone-cloud-kafka-rest |   at org.jvnet.hk2.internal.ServiceLocatorImpl.getUnqualifiedService(ServiceLocatorImpl.java:770)
standalone-cloud-kafka-rest |   at org.jvnet.hk2.internal.IterableProviderImpl.get(IterableProviderImpl.java:87)
standalone-cloud-kafka-rest |   at io.confluent.kafkarest.resources.v3.ProduceAction.recordResponseMetrics(ProduceAction.java:281)
standalone-cloud-kafka-rest |   at io.confluent.kafkarest.resources.v3.ProduceAction.lambda$produce$8(ProduceAction.java:194)
standalone-cloud-kafka-rest |   at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
standalone-cloud-kafka-rest |   at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
standalone-cloud-kafka-rest |   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
standalone-cloud-kafka-rest |   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
standalone-cloud-kafka-rest |   at java.base/java.lang.Thread.run(Thread.java:829)


The record metrics call is done just after we create the ProduceResponse, but before we return it, so I'm a bit surprised this is out of scope of the request.

Reply:
This exception happens because when you call `Provider#get`, that's when the injection happen. For you to be able to inject something `RequestScoped`, you need to be in the request-scope, i.e. running on the request handling thread. If you follow the stack trace, you will notice that you are not in fact running on the request handling thread. You are running it from an async callback.

When do you need a `Provider`? Whenever you need to inject something that is `RequestScoped`. Since injection happens when the constructor of the resource (e.g. `new ProduceAction(...)`) is called, the request scope is not available yet. So injecting a `RequestScoped` thing without a provider in the constructor will do one of two things: 1. inject the wrong thing or 2. fail because it can't even figure out how to inject the wrong thing. That's we use the `Provider`, so that we delay injection until the resource method (e.g. `void produce(...)`) is called, since by then, the request scope is already set.

Why does not `ProducerControllerImpl` inject a Provider of the Producer? Well, because we inject a `Provider<ProducerController>` here in `ProduceAction`. By the time we instantiate `ProducerControllerImpl`, when calling `Provider<ProducerController>#get`, we are already in the request scope, so the right `Producer` is injected. In other words: When `new ProducerControllerImpl` is called, we KNOW we are in the request scope. That's not true for when `new ProducerAction` is called.

What about `ProducerMetrics`? It doesn't have any request scoped state, neither it injects anything request scoped, so it should really be a Singleton, no RequestScoped. I think by not using a Provider, you are injecting the "wrong thing", which turns out to be okay since you never needed a request scoped `ProducerMetrics` anyway.

That would be the end of it if it weren't for your changes in `ce-kafka-rest`. There, we a binding a `RequestScoped` implementation because we need to. The `ProducerMetrics` there needs request scoped information, e.g. the cluster ID, etc. If you try to inject that one here, it will fail because this information is not available. Your tests probably succeeded because in the current PR you are not actually using any request scoped information, but you intend to.

TL;DR: We still need a provider here. Make sure to call `Provider#get` in the request thread (for an example, see where we call the get() for the other providers). Then make sure the ProducerMetrics here in `kafka-rest` is bound as a `Singleton`. This one here doesn't not need to be `RequestScoped`.

Reply:
done